lib/krb5/krcache.c

```diff
@@ -0,0 +1,1994 @@
+ */
+#define KRCC_NAME_PREFIX		"krb_ccache_"
+#define KRCC_NAME_RAND_CHARS		8
+
+#define KRCC_COLLECTION_VERSION		1
+
+#define KRCC_PERSISTENT_ANCHOR		"persistent"
+#define KRCC_PROCESS_ANCHOR		"process"
+#define KRCC_THREAD_ANCHOR		"thread"
+#define KRCC_SESSION_ANCHOR		"session"
+#define KRCC_USER_ANCHOR		"user"
+#define KRCC_LEGACY_ANCHOR		"legacy"
+
+typedef struct _krcc_cursor {
+    size_t numkeys;
+    size_t currkey;
+    key_serial_t princ_id;
+    key_serial_t offsets_id;
+    key_serial_t *keys;
+} *krcc_cursor;
+
+/*
+ * This represents a credentials cache "file" where cache_id is the keyring
+ * serial number for this credentials cache "file".  Each key in the keyring
+ * contains a separate key.
+ */
+typedef struct _krb5_krcache {
+    char *name;		 /* Name for this credentials cache */
+    HEIMDAL_MUTEX lock;	 /* synchronization */
+    key_serial_t collection_id; /* collection containing this cache keyring */
+    key_serial_t cache_id;      /* keyring representing ccache */
+    key_serial_t princ_id;      /* key holding principal info */
+    krb5_timestamp changetime;
+    krb5_boolean is_legacy_type;
+} krb5_krcache;
+
+#define KRCACHE(X) ((krb5_krcache*)(X)->data.data)
+
+static HEIMDAL_MUTEX krcc_mutex = HEIMDAL_MUTEX_INITIALIZER;
+
+static const char *KRB5_CALLCONV
+krcc_get_name(krb5_context context, krb5_ccache id);
+
+static krb5_error_code KRB5_CALLCONV
+_krcc_get_first_internal(krb5_context, krb5_ccache id, krb5_cc_cursor *cursor);
+
+static krb5_error_code KRB5_CALLCONV
+krcc_get_next(krb5_context context,
+	      krb5_ccache id,
+	      krb5_cc_cursor *cursor,
+	      krb5_creds *creds);
+
+static krb5_error_code KRB5_CALLCONV
+krcc_end_get(krb5_context context,
+	     krb5_ccache id,
+	     krb5_cc_cursor *cursor);
+
+static krb5_error_code KRB5_CALLCONV
+krcc_end_cache_get(krb5_context context, krb5_cc_cursor cursor);
+
+static krb5_error_code
+clear_cache_keyring(krb5_context context, krb5_ccache id);
+
+static krb5_error_code
+make_krcache(krb5_context context,
+	     const char *anchor_name,
+	     const char *collection_name,
+	     const char *subsidiary_name,
+	     key_serial_t cache_id,
+	     key_serial_t collection_id,
+	     krb5_krcache **datapp);
+
+static krb5_error_code
+save_principal(krb5_context context, krb5_ccache id, krb5_principal princ);
+
+static krb5_error_code
+save_time_offsets(krb5_context context,
+		  krb5_ccache id,
+		  int32_t sec_offset,
+		  int32_t usec_offset);
+
+static krb5_error_code
+get_time_offsets(krb5_context context,
+		 krb5_ccache id,
+		 int32_t *sec_offset,
+		 int32_t *usec_offset);
+
+static void
+update_change_time(krb5_context context, krb5_krcache *d);
+
+/*
+ * GET_PERSISTENT(uid) acquires the persistent keyring for uid, or falls back
+ * to the user keyring if uid matches the current effective uid.
+ */
+
+static key_serial_t
+get_persistent_fallback(uid_t uid)
+{
+    return (uid == geteuid()) ? KEY_SPEC_USER_KEYRING : -1;
```

Comment:
I don't quite grok why `geteuid()` instead of `getuid()` here.

Reply:
No idea (yet). Perhaps @greghudson can enlighten?